metadatastore/commands.py

```diff
@@ -345,34 +468,69 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
     (not stored, but exchanged) timestamps is optional. If not included,
     data is expected to include (value, timestamp) tuples.
     """
-    if timestamps is not None and set(data.keys()) != set(timestamps.keys()):
-        raise ValueError("The fields in 'data' and 'timestamps' must match.")
-    if timestamps is not None:
-        # new Document exchange format
-        val_ts_tuple = _transform_data(data, timestamps)
-    else:
-        # old Document exchange format
-        val_ts_tuple = data
+    val_ts_tuple = _transform_data(data, timestamps)
 
     # Allow caller to beg forgiveness rather than ask permission w.r.t
     # EventDescriptor creation.
     if descriptor is None:
         raise EventDescriptorIsNoneError()
-
-    if uid is None:
-        uid = str(uuid.uuid4())
-
-    descriptor = _get_mongo_document(descriptor, EventDescriptor)
-    event = Event(descriptor_id=descriptor, uid=uid,
+    descriptor = event_desc_given_uid(descriptor)
+    desc_oid = _EVENTDESC_UID_to_OID_MAP[descriptor['uid']]
+    event = Event(descriptor_id=desc_oid, uid=uid,
                   data=val_ts_tuple, time=time, seq_num=seq_num)
 
-    event = _replace_event_data_key_dots(event, direction='in')
     event.save(validate=True, write_concern={"w": 1})
+
     logger.debug("Inserted Event with uid %s referencing "
                  "EventDescriptor with uid %s", event.uid,
                  descriptor.uid)
     return uid
 
+BAD_KEYS_FMT = """Event documents are malformed, the keys on 'data' and
+'timestamps do not match:\n data: {}\ntimestamps:{}"""
+
+
+@_ensure_connection
+def bulk_insert_events(event_descriptor, events, validate=False):
+    """Bulk insert many events
+
+    Parameters
+    ----------
+    event_descriptor : str
+        The event descriptor uid that these events are associated with
+
+    events : iterable
+       iterable of dicts matching the bs.Event schema
+
+    """
+
+    descriptor = event_desc_given_uid(event_descriptor)
```

Comment:
find_one should be orders of magnitude faster. I think you can squeeze more performance out of this